### ISSUE TITLE
seg fault when creating a new level/snapshot

### ISSUE THREAD

**Issue by whatsthecraic** [the-data-lab/GraphOne#11]
Hi there,

I'm facing a SEGV on a machine, when using multiple threads, even though invocations to #batch_edge are in mutual exclusion, protected by a lock. See attached log. A repro for the current crash is a bit dauting to provide as it will require extracting quite a bit of code, but it seems to occur when the machine is overloaded with more logical threads active than the amount of physical threads.

Let me know if you need more info.

[output.log](https://github.com/the-data-lab/GraphOne/files/3932307/output.log)

**Comment by pradeep-k** [comment]
Without the testcase, it would be hard to understand why the crash happened. 
Does it only occur when the system is overloaded? 
Does a single threaded batch_edge() work fine?

**Comment by whatsthecraic** [comment]
I think a simple repro would be just to invoke batch_edge concurrently from multiple threads. With 1 or 2 threads I haven't hit the crash. On my machine, a relatively old octacore FX-8350, it occurs always when perfoming the updates concurrently with 8 logical threads.

**Comment by pradeep-k** [comment]
Can you tell me the number of edges that are in your dataset? What is the size of property? 64bit or 32bit GraphOne?
What is the size of RAM in your machine?

**Comment by whatsthecraic** [comment]
64-bit, undirected graph, the weights are 64-bit doubles, 16384 vertices,  67092481 edges in total, the vertex array is initialised with 64M vertices. I think the machine is 32GB but I need to double check as it's a different one from what I'm using right now.

**Comment by pradeep-k** [comment]
Thanks for the update. If possible, can you provide a pseudo code, and the data file.  

If the data is a single text file, it would be hard to achieve parallelism for insertion. Do you have data hosted in multiple files?

**Comment by whatsthecraic** [comment]
oky, I'll send you a repro later this evening.

**Comment by whatsthecraic** [comment]
Attached repro: https://gist.github.com/whatsthecraic/b09dd7d5948dc21cf559abbab1c5743f
Compile it with:
```
clang++ -std=c++17 repro.cpp -DOVER_COMMIT -DDEL -DTBB -DPLAIN_GRAPH -DB64 -I<PATH_GRAPHONE>/src -I/home/dean/workspace/graphone/gview -I<PATH_GRAPHONE>/include -I<PATH_GRAPHONE>/onedata -I<PATH_GRAPHONE>/analytics -O0 -g3 -fno-limit-debug-info -Wall  -fopenmp -pthread <PATH_GRAPHONE_BUILD>/src/libsrc64.a <PATH_GRAPHONE_BUILD>/onedata/libonedata64.a -ltbb -o repro.out
```

```
# lldb -- ./repro.out 
(lldb) target create "./repro.out"
Current executable set to './repro.out' (x86_64).
(lldb) r
Process 87329 launched: repro.out' (x86_64)
Generating the stream of edges ... 
Inserting 16384 vertices...
Inserting 67092481 edges with 12 threads ...
corrupted size vs. prev_size
Process 87329 stopped
* thread #24, name = 'repro.out', stop reason = signal SIGABRT
    frame #0: 0x00007ffff797ff25 libc.so.6`__GI_raise + 325
libc.so.6`__GI_raise:
->  0x7ffff797ff25 <+325>: movq   0x108(%rsp), %rax
    0x7ffff797ff2d <+333>: xorq   %fs:0x28, %rax
    0x7ffff797ff36 <+342>: jne    0x7ffff797ff5c            ; <+380>
    0x7ffff797ff38 <+344>: movl   %r8d, %eax
(lldb) bt
* thread #24, name = 'repro.out', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff797ff25 libc.so.6`__GI_raise + 325
    frame #1: 0x00007ffff7969897 libc.so.6`__GI_abort + 299
    frame #2: 0x00007ffff79c3258 libc.so.6`__libc_message + 696
    frame #3: 0x00007ffff79ca77a libc.so.6`malloc_printerr + 26
    frame #4: 0x00007ffff79cb006 libc.so.6`unlink_chunk.isra.0 + 182
    frame #5: 0x00007ffff79cc4bb libc.so.6`_int_free + 1611
    frame #6: 0x0000555555562be5 repro.out`edge_shard_t<dst_weight_t<__univeral_type> >::cleanup(this=0x00005555555af560) at edge_sharding.h:146:13
    frame #7: 0x000055555556165e repro.out`edge_shard_t<dst_weight_t<__univeral_type> >::classify_u(this=0x00005555555af560, pgraph=0x00005555555af2c0) at edge_sharding.h:216:5
    frame #8: 0x000055555555c279 repro.out`::.omp_outlined._debug__(.global_tid.=0x00007ffe56ff9e20, .bound_tid.=0x00007ffe56ff9e18, this=0x00005555555af2c0) at sgraph2.h:59:41
    frame #9: 0x000055555555c2f5 repro.out`::.omp_outlined.(.global_tid.=0x00007ffe56ff9e20, .bound_tid.=0x00007ffe56ff9e18, this=0x00005555555af2c0) at sgraph2.h:57:5
    frame #10: 0x00007ffff7bed763 libomp.so`__kmp_invoke_microtask + 147
    frame #11: 0x00007ffff7b8c91d libomp.so`___lldb_unnamed_symbol279$$libomp.so + 301
    frame #12: 0x00007ffff7b8b92a libomp.so`___lldb_unnamed_symbol268$$libomp.so + 314
    frame #13: 0x00007ffff7bdd302 libomp.so`___lldb_unnamed_symbol869$$libomp.so + 354
    frame #14: 0x00007ffff7b164cf libpthread.so.0`start_thread + 223
    frame #15: 0x00007ffff7a432d3 libc.so.6`__GI___clone + 67
```

On my machine it crashes almost immediately.

**Comment by pradeep-k** [comment]
Test passed.  Commit id is 54e85015809414e9770f9317da468154726505bd
```
pkumar@mira:~/src/GraphOne/build$ ./graphone64
Generating the stream of edges ...
Inserting 16384 vertices...
Inserting 67092481 edges with 20 threads ...
Build ...
Done
```

This is a critical fix, and has not been thoroughly tested. If you see any thing being broken. Please post.

**Comment by whatsthecraic** [comment]
Hi @pradeep-k . Thanks for the quick fix! It improves, still I'm not able to validate the results. Attached a test case extracted from our suite. 

See test: https://gist.github.com/whatsthecraic/5c5fc7c8dcc082a5cf1cd36a955046eb

[log2.txt](https://github.com/the-data-lab/GraphOne/files/3935095/log2.txt)

**Comment by whatsthecraic** [comment]
This assertion is hit  when running the test case for 10 seconds:

```
cpp_hello_graphone: /linux/workspace/graphone/onedata/mem_pool.h:147: status_t thd_mem_t<dst_weight_t<__univeral_type> >::delta_adjlist_bulk(index_t) [T = dst_weight_t<__univeral_type>]: Assertion `mem1->adjlog_beg' failed.

Thread 326 "cpp_hello_graph" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffe3e2c3800 (LWP 23305)]
0x00007ffff796ef25 in raise () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff796ef25 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff7958897 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff7958767 in __assert_fail_base.cold () from /usr/lib/libc.so.6
#3  0x00007ffff7967526 in __assert_fail () from /usr/lib/libc.so.6
#4  0x000055555557417b in thd_mem_t<dst_weight_t<__univeral_type> >::delta_adjlist_bulk (this=0x5555555eacc0, 
    size=18446744073708879872) at /linux/workspace/graphone/onedata/mem_pool.h:147
#5  0x0000555555573fd0 in thd_mem_t<dst_weight_t<__univeral_type> >::alloc_adjlist (this=0x5555555eacc0, 
    count=-42233, hub=true) at /linux/workspace/graphone/onedata/mem_pool.h:108
#6  0x0000555555573dee in onegraph_t<dst_weight_t<__univeral_type> >::new_delta_adjlist (this=0x5555555eabc0, 
    count=-42233, hub=true) at /linux/workspace/graphone/onedata/graph_base.h:143
#7  0x0000555555573802 in onegraph_t<dst_weight_t<__univeral_type> >::del_nebr_noatomic (this=0x5555555eabc0, vid=4, 
    sid=...) at /linux/workspace/graphone/onedata/onegraph.h:186
#8  0x0000555555573256 in onegraph_t<dst_weight_t<__univeral_type> >::archive (this=0x5555555eabc0, 
    edges=0x7ffe38050c30, count=24759, a_snapid=6) at /linux/workspace/graphone/onedata/onegraph.h:37
#9  0x0000555555576248 in edge_shard_t<dst_weight_t<__univeral_type> >::archive (this=0x5555555e8a50, 
    sgraph=0x5555555e8a30, global_range=0x5555555e8aa0, j_start=2, j_end=3, snap_id=5)
    at /linux/workspace/graphone/src/edge_sharding.h:799
#10 0x0000555555574d78 in edge_shard_t<dst_weight_t<__univeral_type> >::classify_u (this=0x5555555e8a50, 
    pgraph=0x5555555e7780) at /linux/workspace/graphone/src/edge_sharding.h:214
#11 0x000055555555efe9 in .omp_outlined._debug__(pgraph_t<dst_weight_t<__univeral_type> > *) const (
    .global_tid.=0x7ffe3e2c2be0, .bound_tid.=0x7ffe3e2c2bd8, this=0x5555555e7780)
    at /linux/workspace/graphone/src/sgraph2.h:59
#12 0x000055555555f065 in .omp_outlined.(void) const (.global_tid.=0x7ffe3e2c2be0, .bound_tid.=0x7ffe3e2c2bd8, 
    this=0x5555555e7780) at /linux/workspace/graphone/src/sgraph2.h:57
#13 0x00007ffff7bba763 in __kmp_invoke_microtask () from /usr/lib/libomp.so
#14 0x00007ffff7b5991d in ?? () from /usr/lib/libomp.so
#15 0x00007ffff7b5892a in ?? () from /usr/lib/libomp.so
#16 0x00007ffff7baa302 in ?? () from /usr/lib/libomp.so
#17 0x00007ffff7f744cf in start_thread () from /usr/lib/libpthread.so.0
#18 0x00007ffff7a322d3 in clone () from /usr/lib/libc.so.6
```

**Comment by whatsthecraic** [comment]
parallel variant: https://gist.github.com/whatsthecraic/3cfd93104fac87670fdefcc1e5bf046f

**Comment by pradeep-k** [comment]
> Hi @pradeep-k . Thanks for the quick fix! It improves, still I'm not able to validate the results. Attached a test case extracted from our suite.
> 
> See test: https://gist.github.com/whatsthecraic/5c5fc7c8dcc082a5cf1cd36a955046eb
> 
> [log2.txt](https://github.com/the-data-lab/GraphOne/files/3935095/log2.txt)

I forgot to mention you last time that the check for eOK after batch_edge() is invalid.  You should check for eOK and eEndBatch both, or you can assume it will take care of everything. There is not error returned from this function.

This makes the testcase success.

**Comment by pradeep-k** [comment]
> This assertion is hit when running the test case for 10 seconds:
> 
> ```
> cpp_hello_graphone: /linux/workspace/graphone/onedata/mem_pool.h:147: status_t thd_mem_t<dst_weight_t<__univeral_type> >::delta_adjlist_bulk(index_t) [T = dst_weight_t<__univeral_type>]: Assertion `mem1->adjlog_beg' failed.
> 
> Thread 326 "cpp_hello_graph" received signal SIGABRT, Aborted.
> [Switching to Thread 0x7ffe3e2c3800 (LWP 23305)]
> 0x00007ffff796ef25 in raise () from /usr/lib/libc.so.6
> (gdb) bt
> #0  0x00007ffff796ef25 in raise () from /usr/lib/libc.so.6
> #1  0x00007ffff7958897 in abort () from /usr/lib/libc.so.6
> #2  0x00007ffff7958767 in __assert_fail_base.cold () from /usr/lib/libc.so.6
> #3  0x00007ffff7967526 in __assert_fail () from /usr/lib/libc.so.6
> #4  0x000055555557417b in thd_mem_t<dst_weight_t<__univeral_type> >::delta_adjlist_bulk (this=0x5555555eacc0, 
>     size=18446744073708879872) at /linux/workspace/graphone/onedata/mem_pool.h:147
> #5  0x0000555555573fd0 in thd_mem_t<dst_weight_t<__univeral_type> >::alloc_adjlist (this=0x5555555eacc0, 
>     count=-42233, hub=true) at /linux/workspace/graphone/onedata/mem_pool.h:108
> #6  0x0000555555573dee in onegraph_t<dst_weight_t<__univeral_type> >::new_delta_adjlist (this=0x5555555eabc0, 
>     count=-42233, hub=true) at /linux/workspace/graphone/onedata/graph_base.h:143
> #7  0x0000555555573802 in onegraph_t<dst_weight_t<__univeral_type> >::del_nebr_noatomic (this=0x5555555eabc0, vid=4, 
>     sid=...) at /linux/workspace/graphone/onedata/onegraph.h:186
> #8  0x0000555555573256 in onegraph_t<dst_weight_t<__univeral_type> >::archive (this=0x5555555eabc0, 
>     edges=0x7ffe38050c30, count=24759, a_snapid=6) at /linux/workspace/graphone/onedata/onegraph.h:37
> #9  0x0000555555576248 in edge_shard_t<dst_weight_t<__univeral_type> >::archive (this=0x5555555e8a50, 
>     sgraph=0x5555555e8a30, global_range=0x5555555e8aa0, j_start=2, j_end=3, snap_id=5)
>     at /linux/workspace/graphone/src/edge_sharding.h:799
> #10 0x0000555555574d78 in edge_shard_t<dst_weight_t<__univeral_type> >::classify_u (this=0x5555555e8a50, 
>     pgraph=0x5555555e7780) at /linux/workspace/graphone/src/edge_sharding.h:214
> #11 0x000055555555efe9 in .omp_outlined._debug__(pgraph_t<dst_weight_t<__univeral_type> > *) const (
>     .global_tid.=0x7ffe3e2c2be0, .bound_tid.=0x7ffe3e2c2bd8, this=0x5555555e7780)
>     at /linux/workspace/graphone/src/sgraph2.h:59
> #12 0x000055555555f065 in .omp_outlined.(void) const (.global_tid.=0x7ffe3e2c2be0, .bound_tid.=0x7ffe3e2c2bd8, 
>     this=0x5555555e7780) at /linux/workspace/graphone/src/sgraph2.h:57
> #13 0x00007ffff7bba763 in __kmp_invoke_microtask () from /usr/lib/libomp.so
> #14 0x00007ffff7b5991d in ?? () from /usr/lib/libomp.so
> #15 0x00007ffff7b5892a in ?? () from /usr/lib/libomp.so
> #16 0x00007ffff7baa302 in ?? () from /usr/lib/libomp.so
> #17 0x00007ffff7f744cf in start_thread () from /usr/lib/libpthread.so.0
> #18 0x00007ffff7a322d3 in clone () from /usr/lib/libc.so.6
> ```

This testcase is little confusing to me. I didn't understand the flow. How many total edges are finally there? How many of them are deleted edges per vertex? The typedef allocates only 16bits for deleted edge count per vertex? 

While I work on it, can you modify and print some log for number of total edges, deleted edges on total, like you did in previous cases.

**Comment by whatsthecraic** [comment]
Hi @pradeep-k, what's the meaning of eEndBatch?

**Comment by whatsthecraic** [comment]
It expects 33 vertices and 90 edges. The thread simply keeps inserting and removing them.
The edges for vertex 1 should be : 2, 3, 4, 5, 6, 7, 8, 9, ... , 32, 33
vertex 2: 1, 2, 4, 6, 8, 10, 12, 14 ..., 32
vertex 3: 1, 3, 6, 9, 12, ... 30, 33
vertex 4: 1, 2, 4, 8, 12, ...., 32
vertex 5: 1, 5, 10... , 30
vertex 6: 1, 2, 6, 12, ..., 30
... so on ...

**Comment by pradeep-k** [comment]
> Hi @pradeep-k, what's the meaning of eEndBatch?

It just means that the system will create a new adjacency list snapshot. Current batching has ended.  Next edge will start a new batch.

**Comment by whatsthecraic** [comment]
> While I work on it, can you modify and print some log for number of total edges, deleted edges on total, like you did in previous cases.

Just to clarify, the assertion is hit  on the same test case, the only difference is:
```
    // let it run for a bit
    LOG("ZzZ ... ");
    this_thread::sleep_for(10s);
```

**Comment by pradeep-k** [comment]
> It expects 33 vertices and 90 edges. The thread simply keeps inserting and removing them.
> The edges for vertex 1 should be : 2, 3, 4, 5, 6, 7, 8, 9, ... , 32, 33
> vertex 2: 1, 4, 6, 8, 10, 12, 14 ..., 32
> vertex 3: 1, 6, 9, 12, ... 30, 33
> vertex 4: 1, 2, 8, 12, ...., 32
> vertex 5: 1, 10... , 30
> vertex 6: 1, 2, 3, 12, ..., 30
> ... so on ...

Due to small number of vertex count, I suspect that the deletion count is going beyond the capacity of 16bits.  Changing the deletion count to 32bits, started printing the validation result. But it throw an error from your code:

```
[DUMP] Vertex: 1, id: 19, degree: 32
[0] destination: 4, id: 25, weight: 220428
[1] destination: 7, id: 7, weight: 691905
[2] destination: 8, id: 8, weight: 36132
[3] destination: 9, id: 0, weight: 557673
[4] destination: 10, id: 24, weight: 461569
[5] destination: 10, id: 24, weight: 733254
[6] destination: 11, id: 2, weight: 633599
[7] destination: 11, id: 2, weight: 379021
[8] destination: 13, id: 1, weight: 60632
[9] destination: 13, id: 1, weight: 62644
[10] destination: 13, id: 1, weight: 132175
[11] destination: 13, id: 1, weight: 55724
[12] destination: 13, id: 1, weight: 559770
[13] destination: 13, id: 1, weight: 147446
[14] destination: 14, id: 9, weight: 622264
[15] destination: 15, id: 30, weight: 796708
[16] destination: 18, id: 20, weight: 305842
[17] destination: 19, id: 27, weight: 654755
[18] destination: 19, id: 27, weight: 339613
[19] destination: 20, id: 14, weight: 653296
[20] destination: 20, id: 14, weight: 232259
[21] destination: 23, id: 18, weight: 482290
[22] destination: 24, id: 3, weight: 603075
[23] destination: 25, id: 31, weight: 773077
[24] destination: 26, id: 32, weight: 286131
[25] destination: 26, id: 32, weight: 222024
[26] destination: 27, id: 28, weight: 268754
[27] destination: 28, id: 15, weight: 627318
[28] destination: 28, id: 15, weight: 680008
[29] destination: 30, id: 11, weight: 84722
[30] destination: 32, id: 29, weight: 315992
[31] destination: 32, id: 29, weight: 379914
[Thread #1478] VALIDATION ERROR (<). Index: 0, Source vertex: 1 [id: 19]. Destination vertex: 4 [id: 25]. Expected: 2
terminate called after throwing an instance of 'char const*'
```

It should be noted that any addition does not check for duplication. We made it that way, though you can customize it. Why the vertices are printing duplicate neighbors? 

I changed the typedef to 32 bits, and removed all the throw statement, and simply print all the output. So, that we can check what it is printing vs what should we expect: 

```
Random seed: 3001321217
Inserting 33 vertices ...
Generating the stream of edges ...
Updating 90 edges with 16 threads ...
[Thread #5468] ZzZ ...
[Thread #5468] Waiting for all threads to terminate ...
Build ...
Validate ...
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 0, Source vertex: 1 [id: 2]. Destination vertex: 7 [id: 21]. Expected: 2
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 1, Source vertex: 1 [id: 2]. Destination vertex: 7 [id: 21]. Expected: 3
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 2, Source vertex: 1 [id: 2]. Destination vertex: 7 [id: 21]. Expected: 4
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 3, Source vertex: 1 [id: 2]. Destination vertex: 10 [id: 12]. Expected: 5
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 4, Source vertex: 1 [id: 2]. Destination vertex: 10 [id: 12]. Expected: 6
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 5, Source vertex: 1 [id: 2]. Destination vertex: 13 [id: 9]. Expected: 7
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 6, Source vertex: 1 [id: 2]. Destination vertex: 13 [id: 9]. Expected: 8
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 7, Source vertex: 1 [id: 2]. Destination vertex: 15 [id: 10]. Expected: 9
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 8, Source vertex: 1 [id: 2]. Destination vertex: 15 [id: 10]. Expected: 10
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 9, Source vertex: 1 [id: 2]. Destination vertex: 15 [id: 10]. Expected: 11
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 10, Source vertex: 1 [id: 2]. Destination vertex: 15 [id: 10]. Expected: 12
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 11, Source vertex: 1 [id: 2]. Destination vertex: 15 [id: 10]. Expected: 13
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 12, Source vertex: 1 [id: 2]. Destination vertex: 17 [id: 31]. Expected: 14
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 13, Source vertex: 1 [id: 2]. Destination vertex: 17 [id: 31]. Expected: 15
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 14, Source vertex: 1 [id: 2]. Destination vertex: 17 [id: 31]. Expected: 16
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 15, Source vertex: 1 [id: 2]. Destination vertex: 20 [id: 26]. Expected: 17
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 16, Source vertex: 1 [id: 2]. Destination vertex: 20 [id: 26]. Expected: 18
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 17, Source vertex: 1 [id: 2]. Destination vertex: 20 [id: 26]. Expected: 19
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
[15] destination: 20, id: 26, weight: 249648
[16] destination: 20, id: 26, weight: 243419
[17] destination: 20, id: 26, weight: 765389
[18] destination: 21, id: 20, weight: 629097
[19] destination: 21, id: 20, weight: 640265
[20] destination: 23, id: 11, weight: 77185
[21] destination: 23, id: 11, weight: 70491
[22] destination: 23, id: 11, weight: 501401
[23] destination: 27, id: 14, weight: 102607
[24] destination: 27, id: 14, weight: 682794
[25] destination: 27, id: 14, weight: 783364
[26] destination: 29, id: 4, weight: 306270
[27] destination: 30, id: 1, weight: 295885
[28] destination: 32, id: 6, weight: 360363
[29] destination: 32, id: 6, weight: 375706
[30] destination: 33, id: 28, weight: 387391
[31] destination: 33, id: 28, weight: 231172
[Thread #5468] VALIDATION ERROR (<). Index: 18, Source vertex: 1 [id: 2]. Destination vertex: 21 [id: 20]. Expected: 20
[DUMP] Vertex: 1, id: 2, degree: 32
[0] destination: 7, id: 21, weight: 583868
[1] destination: 7, id: 21, weight: 268794
[2] destination: 7, id: 21, weight: 788107
[3] destination: 10, id: 12, weight: 141461
[4] destination: 10, id: 12, weight: 691061
[5] destination: 13, id: 9, weight: 131712
[6] destination: 13, id: 9, weight: 290020
[7] destination: 15, id: 10, weight: 30797
[8] destination: 15, id: 10, weight: 210106
[9] destination: 15, id: 10, weight: 579520
[10] destination: 15, id: 10, weight: 768408
[11] destination: 15, id: 10, weight: 41117
[12] destination: 17, id: 31, weight: 459374
[13] destination: 17, id: 31, weight: 399185
[14] destination: 17, id: 31, weight: 746881
..... many more.

```

**Comment by pradeep-k** [comment]
I will leave it to you to point out the problem from the output.  And give me a simple testcase to produce this. May be just a few iteration will be sufficient instead of waiting 10 second.

Also, the nature of testcase will make the 32bit typedef insufficient after some more time. The solution is to run compaction() regularly. Today, it must be done manually by the user. 

We have plans to run it automatically. I have done half changes, but some more are pending.  Right now, I don't have bandwidth to do it.  If you have time, let me know. It will result in a nice research paper.

**Comment by whatsthecraic** [comment]
Hi @pradeep-k ,

there should be no duplicates in the output. For instance consider the [log2.txt above](https://github.com/the-data-lab/GraphOne/files/3935095/log2.txt) 

[DUMP] Vertex: 3, id: 21, degree: 12
[0] destination: 1, id: 18, weight: 355609
[1] destination: 6, id: 11, weight: 179490
[2] destination: 9, id: 31, weight: 671662
[3] destination: 12, id: 6, weight: 588332
[4] destination: 15, id: 14, weight: 190056
[5] destination: 18, id: 22, weight: 683834
[6] destination: 21, id: 24, weight: 266950
[7] destination: 21, id: 24, weight: 609291
[8] destination: 24, id: 23, weight: 343525
[9] destination: 27, id: 3, weight: 712141
[10] destination: 30, id: 25, weight: 663585
[11] destination: 33, id: 9, weight: 579728
[Thread #22304] VALIDATION ERROR (<). Index: 7, Source vertex: 3 [id: 21]. Destination vertex: 21 [id: 24]. Expected: 24

Basically vertex 21 appears twice. Grepping by weight 266950 and  609291 reports that edge 3 -> 21 [609291] appears at line 194, and at line 374 for weight 609291. 

Againg grepping by 3 -> 21 tells where the edge has been inserted:
```
# grep -Inr "3 -> 21" log2.txt
14:[Thread #22628] INSERT 3 -> 21 [21 -> 24, weight = 303260]
104:[Thread #22628] REMOVE 3 -> 21 [21 -> 24, weight = 0]
....
167234:[Thread #22628] INSERT 3 -> 21 [21 -> 24, weight = 703648]
167324:[Thread #22628] REMOVE 3 -> 21 [21 -> 24, weight = 0]
167414:[Thread #22628] INSERT 3 -> 21 [21 -> 24, weight = 276233]
167504:[Thread #22628] REMOVE 3 -> 21 [21 -> 24, weight = 0]
167594:[Thread #22628] INSERT 3 -> 21 [21 -> 24, weight = 634331]
167684:[Thread #22628] REMOVE 3 -> 21 [21 -> 24, weight = 0]
167774:[Thread #22628] INSERT 3 -> 21 [21 -> 24, weight = 554732]
```

Basically 3 -> 21 should only appear once, and its weight should be 554732

**Comment by pradeep-k** [comment]
Thanks for the update. Fix is done.  Here is the output:
```
pkumar@mira:~/src/GraphOne/build$ ./graphone64
Random seed: 1590332359
Inserting 33 vertices ...
Generating the stream of edges ...
Updating 90 edges with 16 threads ...
[Thread #16382] ZzZ ...
[Thread #16382] Waiting for all threads to terminate ...
Build ...
Validate ...
Done
```

You need to change the 16bit type to 32 bit in here yourself:
```
class sdegree_t {
 public:
    degree_t add_count;
    uint16_t del_count;  //Change this typedef in onedata/vunit.h

```

As I said, this change is temporary, as compaction need to run periodically.  

Fix commit is 8e10fdcef02c58c13ec8900ad218e9e04778d69d

**Comment by whatsthecraic** [comment]
Hi @pradeep-k ,
again, many thanks for the quick fix. I am aware these test cases are starting being more and more complicated and exaggerated. My point is to run the whole system for several hours. Eventually, all results produced should match a given reference. The aim of these tests is to try to catch any variation as soon as possible, rather than finding a mismatch, occured at whatever point, when retrieving the results after a long execution..

A few questions about compactation:
1. Does it imply merging all deltas/changes and rebuilding the whole internal data structure? 
2. The API to use is `p_ugraph_t::compress_graph_baseline()` ?
3. Is it thread safe ? Or should it be protected within an external lock?
4. How often does it need to be invoked? After N updates, N global deletions, N deletions referring the same node?

Again, thanks for the quick info and the provided patches.

**Comment by pradeep-k** [comment]
I am working on it. Can you raise another issue regarding compaction?

**Comment by pradeep-k** [comment]
You can call `p_ugraph_t::compress_graph_baseline()` for compacting and removing the deleted edges.  

Current implementation does not allow compaction to be done while either  (or both) archiving or reading is being done.  It can run in parallel to logging.  Hence, it should ideally be called from archiving. I have modified your testcase to call compact(). I am sure we can do better placement of compact, but need some effort.

Our final aim is to call compaction from archive, and allow parallel reads. But as I told you we are not there. There will be some race conditions in reading if we do that today.

**Comment by pradeep-k** [comment]
Commit number is 8721174449eee6fca385198088183d731e159044 
Kindly use -DDEL in your compilation flag for testing.

**Comment by pradeep-k** [comment]
Answers are as per current status:

- Does it imply merging all deltas/changes and rebuilding the whole internal data structure? It implies changes in  edge arrays of each affected vertices, and updating the degree of affected vertices.

-The API to use is p_ugraph_t::compress_graph_baseline() ? Yes
-Is it thread safe ? Or should it be protected within an external lock? It need to be called in controlled environment, akin to a lock.
-How often does it need to be invoked? After N updates, N global deletions, N deletions referring the same node? At the extreme is when deletion degree runs out of range. I have put an assert to warn the user when they run out. 

`Plan:` The plan is call them automatically and in incremental fashion, i.e. a compacting only a few vertices at a time. The trigger could be specified by the user, such as when the deletion count per vertex is greater than X. The compaction code should get triggered when a vertex achieves X.

**Comment by whatsthecraic** [comment]
oky, thanks for the heads-up. I confirm the patch works. Thanks, Dean

**Comment by pradeep-k** [comment]
Thanks for the confirmation.
I am working on automatic call of compaction as per the above plan. I have seen some initial success. A few more things are pending, before I confirm the working code.

**Comment by pradeep-k** [comment]
Feature branch contains automatic compaction now. It takes care of all the race-conditions.  Hence, any parallel operations can take place while compaction happening internally including parallel read and/or parallel batch_edge.  

The code is very complex, and is part of an upcoming research paper (in draft phase). If you find any issues, do not hesitate to create a new issue.  I am closing this bug.
